app/post_metadata.go

```diff
@@ -0,0 +1,418 @@
+
+func (a *App) InitPostMetadata() {
+	// Dump any cached links if the proxy settings have changed so image URLs can be updated
+	a.AddConfigListener(func(before, after *model.Config) {
+		if (before.ServiceSettings.ImageProxyType != after.ServiceSettings.ImageProxyType) ||
+			(before.ServiceSettings.ImageProxyURL != after.ServiceSettings.ImageProxyType) {
+			linkCache.Purge()
+		}
+	})
+}
+
+func (a *App) PreparePostListForClient(originalList *model.PostList) *model.PostList {
+	numPosts := len(originalList.Posts)
+
+	list := &model.PostList{
+		Posts: make(map[string]*model.Post, numPosts),
+		Order: originalList.Order, // Note that this uses the original Order array, so it isn't a deep copy
+	}
+
+	posts := make(chan *model.Post, numPosts)
+
+	for _, originalPost := range originalList.Posts {
+		go func(originalPost *model.Post) {
+			posts <- a.PreparePostForClient(originalPost)
+		}(originalPost)
+	}
+
+	for i := 0; i < numPosts; i++ {
+		post := <-posts
+		list.Posts[post.Id] = post
+	}
+
+	close(posts)
+
+	return list
+}
+
+func (a *App) PreparePostForClient(originalPost *model.Post) *model.Post {
+	post := originalPost.Clone()
+
+	// Proxy image links before constructing metadata so that requests go through the proxy
+	post = a.PostWithProxyAddedToImageURLs(post)
+
+	if post.Metadata == nil {
+		post.Metadata = &model.PostMetadata{}
+
+		// Emojis and reaction counts
+		if emojis, reactions, err := a.getEmojisAndReactionsForPost(post); err != nil {
+			mlog.Warn("Failed to get emojis and reactions for a post", mlog.String("post_id", post.Id), mlog.Any("err", err))
+		} else {
+			post.Metadata.Emojis = emojis
+			post.Metadata.Reactions = reactions
+		}
+
+		// Files
+		if fileInfos, err := a.getFileMetadataForPost(post); err != nil {
+			mlog.Warn("Failed to get files for a post", mlog.String("post_id", post.Id), mlog.Any("err", err))
+		} else {
+			post.Metadata.Files = fileInfos
+		}
+
+		// Embeds and image dimensions
+		firstLink, images := getFirstLinkAndImages(post.Message)
+
+		if embed, err := a.getEmbedForPost(post, firstLink); err != nil {
+			mlog.Warn("Failed to get embedded content for a post", mlog.String("post_id", post.Id), mlog.Any("err", err))
+		} else if embed == nil {
+			post.Metadata.Embeds = []*model.PostEmbed{}
+		} else {
+			post.Metadata.Embeds = []*model.PostEmbed{embed}
+		}
+
+		post.Metadata.Images = a.getImagesForPost(post, images)
+	}
+
+	return post
+}
+
+func (a *App) getFileMetadataForPost(post *model.Post) ([]*model.FileInfo, *model.AppError) {
+	if len(post.FileIds) == 0 { // This field is deprecated, but still use it for now to avoid unnecessary database hits
+		return nil, nil
+	}
+
+	return a.GetFileInfosForPost(post.Id, false)
+}
+
+func (a *App) getEmojisAndReactionsForPost(post *model.Post) ([]*model.Emoji, []*model.Reaction, *model.AppError) {
+	var reactions []*model.Reaction
+	if post.HasReactions { // This field is deprecated, but still use it for now to avoid unnecessary database hits
```

Comment:
Maybe I'll remove these comments since I'm undecided on whether or not we'll remove these fields. The reason I wanted to deprecate them is because they represent a two-way linkage between Posts and Reactions/Files making them harder to keep track of, and they were originally intended to combat scroll pop before post metadata was added (although I think Post.HasReactions got broken). At the same time, they're really convenient here since they let us avoid the extra database hit just as they were used before to avoid the extra API call prior to this.

I think I'll remove these deprecation comments and make a note to take another look at them later once all this has settled